File: DelSole.CrossAnalyzers.VisualBasic/DelSole.CrossAnalyzers.VisualBasic/ObservableCollectionAnalyzer.vb
<DiagnosticAnalyzer(LanguageNames.VisualBasic)>
Public Class DelSoleCrossAnalyzersVisualBasicAnalyzer
    Inherits DiagnosticAnalyzer

    Public Const DiagnosticId As String = "CRA012"
    Friend Shared ReadOnly Title As String = "List(Of T) is improper for data-binding"
    Friend Shared ReadOnly MessageFormat As String = "'{0}' is of type List(Of T). Consider assigning an object of type ObservableCollection(Of T) instead."
    Friend Shared ReadOnly Category As String = "Platform"

    Private Shared Rule As New DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, isEnabledByDefault:=True,
                                                    helpLinkUri:="https://github.com/AlessandroDelSole/DelSole.CrossAnalyzers/wiki")

    Public Overrides ReadOnly Property SupportedDiagnostics As ImmutableArray(Of DiagnosticDescriptor)
        Get
            Return ImmutableArray.Create(Rule)
        End Get
    End Property

    Public Overrides Sub Initialize(context As AnalysisContext)
        ' TODO: Consider registering other actions that act on syntax instead of or in addition to symbols
        ' See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Analyzer%20Actions%20Semantics.md for more information
        context.RegisterSyntaxNodeAction(AddressOf AnalyzeAssignementStatements, SyntaxKind.SimpleAssignmentStatement)
    End Sub

    Private Sub AnalyzeSymbol(context As SymbolAnalysisContext)
        ' TODO: Replace the following code with your own analysis, generating Diagnostic objects for any issues you find

        Dim namedTypeSymbol = CType(context.Symbol, INamedTypeSymbol)

        ' Find just those named type symbols with names containing lowercase letters.
        If namedTypeSymbol.Name.ToCharArray.Any(AddressOf Char.IsLower) Then
            ' For all such symbols, produce a diagnostic.
            Dim diag = Diagnostic.Create(Rule, namedTypeSymbol.Locations(0), namedTypeSymbol.Name)

            context.ReportDiagnostic(diag)
        End If
    End Sub

    Public Sub AnalyzeAssignementStatements(context As SyntaxNodeAnalysisContext)
        Dim node = TryCast(context.Node, AssignmentStatementSyntax)
        If node Is Nothing Then
            Return
        End If

        Dim leftPart = node.Left.ToFullString
        If leftPart.Contains("ItemsSource") = False Then
            If leftPart.Contains("DataContext") = False Then
                Return
            End If
        End If

        Dim returnType = ""
        Dim SymbolInfo = context.SemanticModel.GetSymbolInfo(node.Right).Symbol
        If TypeOf (SymbolInfo) Is ILocalSymbol Then
            returnType = CType(SymbolInfo, ILocalSymbol).Type.ToString()
        ElseIf TypeOf (SymbolInfo) Is IPropertySymbol Then
            returnType = CType(SymbolInfo, IPropertySymbol).Type.ToString()
        ElseIf TypeOf (SymbolInfo) Is IFieldSymbol Then
            returnType = CType(SymbolInfo, IFieldSymbol).Type.ToString()
        ElseIf TypeOf (SymbolInfo) Is IMethodSymbol Then
            returnType = CType(SymbolInfo, IMethodSymbol).ReturnType.ToString()
        Else
            Return
        End If

        If Not returnType.ToLowerInvariant().Contains("system.collections.generic.list") Then Return

        Dim diagn = Diagnostic.Create(Rule, node.Right.GetLocation(), node.Right.ToString())
        context.ReportDiagnostic(diagn)
    End Sub
End Class

File: DelSole.CrossAnalyzers.CSharp/DelSole.CrossAnalyzers.CSharp/ObservableCollectionAnalyzer.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Diagnostics;

namespace DelSole.CrossAnalyzers.CSharp
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class ObservableCollectionAnalyzer : DiagnosticAnalyzer
    {
        public const string DiagnosticId = "CRA001";
        internal static readonly string Title = "List<T> is improper for data-binding";
        internal static readonly string MessageFormat = "'{0}' is of type List<T>. Consider assigning an object of type ObservableCollection<T> instead.";
        internal const string Category = "Platform";

        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticId, Title, MessageFormat, Category, DiagnosticSeverity.Warning, true, helpLinkUri: "https://github.com/AlessandroDelSole/DelSole.CrossAnalyzers/wiki");

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterSyntaxNodeAction(AnalyzeAssignementStatements, SyntaxKind.SimpleAssignmentExpression);
        }


        public void AnalyzeAssignementStatements(SyntaxNodeAnalysisContext context)
        {
                var node = context.Node as AssignmentExpressionSyntax;
                if (node == null)
                {
                    return;
                }

                //Check if the node contains bindable properties
                var leftPart = node.Left.ToFullString();
                if (leftPart.Contains("ItemsSource") || leftPart.Contains("DataContext") || leftPart.Contains("BindingContext"))
                {

                string returnType = "";
                var symbolInfo = context.SemanticModel.GetSymbolInfo(node.Right).Symbol;
                //Check if local variable
                if ((symbolInfo) is ILocalSymbol)
                {
                    returnType = ((ILocalSymbol)symbolInfo).Type.ToString();
                    //Check if property
                }
                else if ((symbolInfo) is IPropertySymbol)
                {
                    returnType = ((IPropertySymbol)symbolInfo).Type.ToString();
                    //Check if field
                }
                else if ((symbolInfo) is IFieldSymbol)
                {
                    returnType = ((IFieldSymbol)symbolInfo).Type.ToString();
                }
                else if ((symbolInfo) is IMethodSymbol)
                {
                    //Check if method
                    returnType = ((IMethodSymbol)symbolInfo).ReturnType.ToString();
                }
                else
                {
                    return;
                }

                if (returnType.ToLowerInvariant().Contains("system.collections.generic.list"))
                {
                    var diagn = Diagnostic.Create(Rule, node.Right.GetLocation(), node.Right.ToString());
                    context.ReportDiagnostic(diagn);
                }
            }

       }
    }
}
